src/WorkerHandler.js

```diff
@@ -377,23 +377,38 @@ WorkerHandler.prototype.terminate = function (force, callback) {
   }
   if (!this.busy()) {
     // all tasks are finished. kill the worker
+    var cleanup = function(err) {
+      me.terminated = true;
+      me.worker = null;
+      me.terminating = false;
+      if (me.terminationHandler) {
+        me.terminationHandler(err, me);
+      } else if (err) {
+        throw err;
+      }
+    }
+
     if (this.worker) {
       if (typeof this.worker.kill === 'function') {
-        this.worker.kill();  // child process
+        // child process
+        // cleanup once the child process has exited
+        this.worker.once('exit', function() {
+          cleanup();
+        });
+        if (!this.worker.killed && !this.worker.kill()) {
+          cleanup(new Error('Failed to send SIGTERM to worker'));
+        }
+        return;
       }
       else if (typeof this.worker.terminate === 'function') {
         this.worker.terminate(); // web worker
+        this.worker.killed = true;
       }
       else {
         throw new Error('Failed to terminate worker');
       }
-      this.worker = null;
-    }
-    this.terminating = false;
-    this.terminated = true;
-    if (this.terminationHandler) {
-      this.terminationHandler(this);
     }
+    cleanup();
```

Comment:
It looks like `cleanup()` is called twice: one time right away, and one time on callback of `this.worker.once('exit'`. Should we split this in two functions? Maybe set `this.terminating = true` when starting to terminate, and `this.terminated = true` after termination finished?

Reply:
It's not called twice unless somehow the child process has already been killed, then emits an additional `exit` event.  I have no idea if this is possible:

https://github.com/josdejong/workerpool/blob/57e6795ca08d51022847f8de204fca7ae0fb8407/src/WorkerHandler.js#L392-L401

However, that should probably change to:

```js
      if (typeof this.worker.kill === 'function') {
        // child process
        if (!this.worker.killed && !this.worker.kill()) {
          cleanup(new Error('Failed to send SIGTERM to worker'));
        } else {          
          // cleanup once the child process has exited
          this.worker.once('exit', function() {
            cleanup();
          });
        }
        return;
```

Reply:
FWIW I'm not entirely confident that just listening for the `exit` event is enough.  Maybe a strategy like [signal-exit](https://npm.im/signal-exit) would be more robust--_if_ that's possible with a child process.

Reply:
👍 I think you're right